### ISSUE TITLE
Feature Request: more strict type inference on constructed "mutable" containers of literals with `in` operator

### ISSUE THREAD

**Issue by Andrej730** [astral-sh/ty#4026]
Hi! Wanted to share an idea I had going through bunch of type narrowing cases ensuring type checker understands them.

As everyone knows `["A", "B"]` is typically inferred as `list[str]`, since list is mutable and it's unsafe to consider that all elements will be ` Literal["A", "B"]`. Same for sets.

<details>
<summary>
current behaviour
</summary>

```python
from typing import Literal, reveal_type

Value = Literal["A", "B", "C"]


def test_list(value: Value):
    if value in ["A", "B"]:
        print(value)
    else:
        # Literal["A", "B", "C"] - not narrowed
        reveal_type(value)


def test_set(value: Value):
    if value in {"A", "B"}:
        print(value)
    else:
        # Literal["A", "B", "C"] - not narrowed
        reveal_type(value)


def test_tuple(value: Value):
    if value in ("A", "B"):
        print(value)
    else:
        # Literal["C"] - correctly narrowed
        reveal_type(value)
```

</details>

It becomes a problem when using `value in [a, b]` and it couldn't type narrow type of `value`, producing all kind of typing issues and doesn't allowing to exhaust original set of literals and add nice `else: assert_never(value)` at the end. As a workaround user can either check values one by one `value == a or value == b` or can use tuples `value in (a, b)`.

Tuples seem as one of the most correct ways to do it - lists mutability isn't needed here, using sets with hashes may not seem worth it for a few items, but many pythonistas just use lists all the time for such cases and some use sets for small performance gains (e.g. there's a ruff rule to convert all such cases to sets - https://docs.astral.sh/ruff/rules/literal-membership/).

So it may be reasonable to allow it? To infer `list[Literal["A", "B"]]` / `set[Literal["A", "B"]]` when the list/set constructor is used just for checking `xxx in [a, b]` / `xxx in {a, b}`, where all container elements are just literals. Thoughts?

[Playground](https://play.ty.dev/ab035aa9-c7c5-4051-a264-12c89aa0185a) with current behaviour
[pyright-playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAySMApiAIYA2ANFCCQG4lUD68CJAUJwGpUCuJKAF5CxMlQDaAIgCC02tIBCCqNIDC0gLrdOAExLBYJAM4wWlJGYAUDASQBcUPpUEBKB5yjfMRu66FUKBl5RRUtTx8oqAQQVBhbezcvHxJKE0cU6KgAYjFSCkoQ1WUSzS0oAFooFDB8FHIQcAB3Ej0s6PomVnYSRIDk7gMjUjMWDIT-QScXd0ifJD97TBQoAG85EpUAX3ns2Pj%2B9w60jL3ovKICqU2wsu0qmrqaxpa2jqiu5ko2RD6pkiDfSGYxjGD8BCUf72GZJc6LKAAlZQay3NQqDwfHwHFCTJIndKZbLeS7iQoycqPADGuHoVJglDgLyaYFa7WJdEY31%2BHCOgM4QA) (doesn't support it too)

PS Btw, fun fact - when doing `xxx in [a, b]`, Python's bytecode actually never construct a list, it just loads a presaved tuple. Same for sets - they're loaded as frozensets. Anyways, it's probably just an implementation detail and probably have no effect on the type checker.

```python
--- test_list ---
  6           RESUME                   0

  7           LOAD_FAST_BORROW         0 (value)
              LOAD_CONST               2 (('A', 'B'))
              CONTAINS_OP              0 (in)
              POP_JUMP_IF_FALSE       14 (to L1)
              NOT_TAKEN
```

Could be a bit related:
- https://github.com/astral-sh/ty/issues/2441
- https://github.com/astral-sh/ty/issues/2280

**Comment by carljm** [comment]
Yes, I think it would make sense to do this. We already do this for iteration (#2441); this is just extending the same fix to the membership-testing case.

One reason to prioritize this is that Ruff PLR6201 explicitly rewrites `x in ("a", "b")` to `x in {"a", "b"}` for performance reasons, which then loses narrowing in ty.

**Comment by Andrej730** [comment]
That was fast, thank you! 
Can confirm it's working - it just saved me adding bunch of noise diffs converting those lists of literals to tuples.

Btw just tested other type checkers support for this:
- no support - pyright, mypy, zuban, basilisk
- supported - ty, pyrefly
